File: android/app/src/main/java/com/retroplay/RNGameManager.java
package com.retroplay;


import android.content.Context;


import android.view.Choreographer;
import android.view.Display;
import android.view.KeyEvent;
import android.view.LayoutInflater;

import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import androidx.fragment.app.FragmentActivity;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.common.MapBuilder;

import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactPropGroup;
import com.swordfish.libretrodroid.GLRetroView;
//import com.swordfish.libretrodroid.GLRetroView;
//import com.swordfish.libretrodroid.GLRetroViewData;


import java.util.Map;

public class RNGameManager extends ViewGroupManager<FrameLayout> {

    public static final String REACT_CLASS = "RNGameManager";
    private static RNGameManager instance = null;
    public final int COMMAND_CREATE = 1;
    ReactApplicationContext mCallerContext;
    private int propWidth;
    private int propHeight;


    public RNGameManager(ReactApplicationContext reactContext) {
        mCallerContext = reactContext;


    }

    public static RNGameManager getInstance() {
        return instance;
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }


    @Override
    public FrameLayout createViewInstance(ThemedReactContext context) {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        FrameLayout frameLayout = (FrameLayout) inflater.inflate(R.layout.game_fragment, null);

        return frameLayout;
    }


    /**
     * Map the "captureImage" command to an integer
     */
    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        return MapBuilder.of("create", COMMAND_CREATE);
    }

    /**
     * Handle "captureImage" command called from JS
     */
    @Override
    public void receiveCommand(@NonNull FrameLayout root, String commandId, @Nullable ReadableArray args) {
        super.receiveCommand(root, commandId, args);
        int reactNativeViewId = args.getInt(0);
        String romID = args.getString(1);
        String coreName = args.getString(2);
        int commandIdInt = Integer.parseInt(commandId);

        switch (commandIdInt) {
            case COMMAND_CREATE:
                createFragment(root, reactNativeViewId, romID, coreName);
                break;
            default: {
            }
        }
    }

    @ReactPropGroup(names = {"width", "height"}, customType = "Style")
    public void setStyle(FrameLayout view, int index, Integer value) {
        if (index == 0) {
            propWidth = value;
        }

        if (index == 1) {
            propHeight = value;
        }
    }

    /**
     * Replace your React Native view with a custom fragment
     */
    public void createFragment(FrameLayout root, int reactNativeViewId, String romID, String coreName) {
        ViewGroup parentView = (ViewGroup) root.findViewById(reactNativeViewId);
        setupLayout(parentView);

        final GameFragment gameFragment = new GameFragment(romID, coreName);
        FragmentActivity activity = (FragmentActivity) mCallerContext.getCurrentActivity();
        activity.getSupportFragmentManager()
                .beginTransaction()
                .replace(reactNativeViewId, gameFragment, String.valueOf(reactNativeViewId))
                .commit();
    }

    public void setupLayout(View view) {
        Choreographer.getInstance().postFrameCallback(new Choreographer.FrameCallback() {
            @Override
            public void doFrame(long frameTimeNanos) {
                manuallyLayoutChildren(view);
                view.getViewTreeObserver().dispatchOnGlobalLayout();
                Choreographer.getInstance().postFrameCallback(this);
            }
        });
    }

    /**
     * Layout all children properly
     */
    public void manuallyLayoutChildren(View view) {
        // propWidth and propHeight coming from react-native props
        int width = 500;
        int height = 500;


        Display metrics = mCallerContext.getCurrentActivity().getWindowManager().getDefaultDisplay();

        view.measure(
                View.MeasureSpec.makeMeasureSpec(metrics.getWidth(), View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(metrics.getHeight(), View.MeasureSpec.EXACTLY));

        view.layout(0, 0, metrics.getWidth(), metrics.getHeight());
        //view.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
    }


    public void onKeyEventDown(int keyCode, KeyEvent event) {
        GLRetroView game_view = (GLRetroView) mCallerContext.getCurrentActivity().findViewById(R.id.gl_retro_view);
        if (game_view != null) {
            game_view.sendKeyEvent(event.getAction(), keyCode, 0);
        }
    }

    public void onKeyEventUp(int keyCode, KeyEvent event) {
        GLRetroView game_view = (GLRetroView) mCallerContext.getCurrentActivity().findViewById(R.id.gl_retro_view);
        if (game_view != null) {
            game_view.sendKeyEvent(event.getAction(), keyCode, 0);
        }
    }

    public void onGenericMotionEvent(MotionEvent event) {
        if (event != null) {
            sendMotionEvent(
                    event,
                    GLRetroView.MOTION_SOURCE_DPAD,
                    MotionEvent.AXIS_HAT_X,
                    MotionEvent.AXIS_HAT_Y,
                    0
            );
            sendMotionEvent(
                    event,
                    GLRetroView.MOTION_SOURCE_ANALOG_LEFT,
                    MotionEvent.AXIS_X,
                    MotionEvent.AXIS_Y,
                    0
            );
            sendMotionEvent(
                    event,
                    GLRetroView.MOTION_SOURCE_ANALOG_RIGHT,
                    MotionEvent.AXIS_Z,
                    MotionEvent.AXIS_RZ,
                    0
            );
        }
    }

    private void sendMotionEvent(MotionEvent event,int source,int xAxis,int yAxis,int port){
        GLRetroView game_view = (GLRetroView) mCallerContext.getCurrentActivity().findViewById(R.id.gl_retro_view);
        if (game_view != null){
            game_view.sendMotionEvent(source,event.getAxisValue(xAxis),event.getAxisValue(yAxis),port);
        }
    }
}

File: android/app/src/main/java/com/retroplay/RNGameManager.kt
package com.retroplay

import android.content.Context
import android.view.*

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ViewGroupManager
import android.widget.FrameLayout
import com.retroplay.RNGameManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.annotations.ReactPropGroup
import com.retroplay.GameFragment
import androidx.fragment.app.FragmentActivity
import com.facebook.react.common.MapBuilder
import com.swordfish.libretrodroid.GLRetroView

//import com.swordfish.libretrodroid.GLRetroView;
//import com.swordfish.libretrodroid.GLRetroViewData;
class RNGameManager(var mCallerContext: ReactApplicationContext) : ViewGroupManager<FrameLayout>() {
    val COMMAND_CREATE = 1
    private var propWidth = 0
    private var propHeight = 0
    override fun getName(): String {
        return REACT_CLASS
    }

    public override fun createViewInstance(context: ThemedReactContext): FrameLayout {
        val inflater =
            context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        return inflater.inflate(R.layout.game_fragment, null) as FrameLayout
    }

    /**
     * Map the "captureImage" command to an integer
     */
    override fun getCommandsMap(): Map<String, Int>? {
        return MapBuilder.of("create", COMMAND_CREATE)
    }

    /**
     * Handle "captureImage" command called from JS
     */
    override fun receiveCommand(root: FrameLayout, commandId: String, args: ReadableArray?) {
        super.receiveCommand(root, commandId, args)
        val reactNativeViewId = args!!.getInt(0)
        val romID = args.getString(1)
        val coreName = args.getString(2)
        val commandIdInt = commandId.toInt()
        when (commandIdInt) {
            COMMAND_CREATE -> createFragment(root, reactNativeViewId, romID, coreName)
            else -> {}
        }
    }

    @ReactPropGroup(names = ["width", "height"], customType = "Style")
    fun setStyle(view: FrameLayout?, index: Int, value: Int) {
        if (index == 0) {
            propWidth = value
        }
        if (index == 1) {
            propHeight = value
        }
    }

    /**
     * Replace your React Native view with a custom fragment
     */
    fun createFragment(
        root: FrameLayout,
        reactNativeViewId: Int,
        romID: String?,
        coreName: String?
    ) {
        val parentView = root.findViewById<View>(reactNativeViewId) as ViewGroup
        setupLayout(parentView)
        val gameFragment = GameFragment(romID, coreName)
        val activity = mCallerContext.currentActivity as FragmentActivity?
        activity!!.supportFragmentManager
            .beginTransaction()
            .replace(reactNativeViewId, gameFragment, reactNativeViewId.toString())
            .commit()
    }

    fun setupLayout(view: View) {
        Choreographer.getInstance().postFrameCallback(object : Choreographer.FrameCallback {
            override fun doFrame(frameTimeNanos: Long) {
                manuallyLayoutChildren(view)
                view.viewTreeObserver.dispatchOnGlobalLayout()
                Choreographer.getInstance().postFrameCallback(this)
            }
        })
    }

    /**
     * Layout all children properly
     */
    fun manuallyLayoutChildren(view: View) {
        // propWidth and propHeight coming from react-native props
        val width = 500
        val height = 500
        val metrics = mCallerContext.currentActivity!!.windowManager.defaultDisplay
        view.measure(
            View.MeasureSpec.makeMeasureSpec(metrics.width, View.MeasureSpec.EXACTLY),
            View.MeasureSpec.makeMeasureSpec(metrics.height, View.MeasureSpec.EXACTLY)
        )
        view.layout(0, 0, metrics.width, metrics.height)
        //view.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
    }

    fun onKeyEventDown(keyCode: Int, event: KeyEvent) {
        val game_view =
            mCallerContext.currentActivity!!.findViewById<View>(R.id.gl_retro_view) as GLRetroView
        game_view?.sendKeyEvent(event.action, keyCode, 0)
    }

    fun onKeyEventUp(keyCode: Int, event: KeyEvent) {
        val game_view =
            mCallerContext.currentActivity!!.findViewById<View>(R.id.gl_retro_view) as GLRetroView
        game_view?.sendKeyEvent(event.action, keyCode, 0)
    }

    fun onGenericMotionEvent(event: MotionEvent?) {
        if (event != null) {
            sendMotionEvent(
                event,
                GLRetroView.MOTION_SOURCE_DPAD,
                MotionEvent.AXIS_HAT_X,
                MotionEvent.AXIS_HAT_Y,
                0
            )
            sendMotionEvent(
                event,
                GLRetroView.MOTION_SOURCE_ANALOG_LEFT,
                MotionEvent.AXIS_X,
                MotionEvent.AXIS_Y,
                0
            )
            sendMotionEvent(
                event,
                GLRetroView.MOTION_SOURCE_ANALOG_RIGHT,
                MotionEvent.AXIS_Z,
                MotionEvent.AXIS_RZ,
                0
            )
        }
    }

    private fun sendMotionEvent(
        event: MotionEvent,
        source: Int,
        xAxis: Int,
        yAxis: Int,
        port: Int
    ) {
        val game_view =
            mCallerContext.currentActivity!!.findViewById<View>(R.id.gl_retro_view) as GLRetroView
        game_view?.sendMotionEvent(
            source,
            event.getAxisValue(xAxis),
            event.getAxisValue(yAxis),
            port
        )
    }

    companion object {
        const val REACT_CLASS = "RNGameManager"
        @JvmStatic
        val instance: RNGameManager? = null
    }
}